p2p/lightchain.py

```diff
@@ -210,19 +224,62 @@ def peer_block_height(peer: LESPeer) -> int:
             return
 
         start_block = await self.get_sync_start_block(peer, head_info)
-        while start_block < head_info.block_number:
+        while start_block < head_info.block_number and peer.connected:
             try:
                 # We should use "start_block + 1" here, but we always re-fetch the last synced
                 # block to work around https://github.com/ethereum/go-ethereum/issues/15447
                 batch = await self.fetch_headers(start_block, peer)
             except TooManyTimeouts:
                 raise LESAnnouncementProcessingError(
                     "Too many timeouts when fetching headers from {}".format(peer))
-            for header in batch:
-                await self.headerdb.coro_persist_header(header)
-                start_block = header.block_number
+            new_tip = await self._import_headers_from_peer(batch, peer)
+            if new_tip is not None:
+                start_block = new_tip
             self.logger.info("synced headers up to #%s", start_block)
 
+    async def _import_headers_from_peer(self, headers, peer):
+        """
+        :return: the block number of the new tip after importing the header
+        """
+        new_tip = None
+        for header in headers:
+            try:
+                await self._validate_header(header)
+            except ValidationError as exc:
+                self.logger.info(
+                    "Dropping Peer %s for sending invalid header %s: %s",
+                    peer,
+                    header,
+                    exc,
+                )
+                # TODO self.peer_pool.blacklist_peer(peer)
+                peer.disconnect(DisconnectReason.bad_protocol)
+                break
+            else:
+                await self.headerdb.coro_persist_header(header)
+                new_tip = header.block_number
+        return new_tip
+
+    async def _validate_header(self, header):
+        if header.is_genesis:
+            parent_header = None
+        else:
+            parent_header = await self._get_parent_header(header)
+        VM = self.Chain.get_vm_class_for_block_number(header.block_number)
+        # TODO push validation into process pool executor
+        VM.validate_header(header, parent_header)
+
+    async def _get_parent_header(self, header):
+        parent_hash = header.parent_hash
+        # TODO extract 500 to constant
+        for attempt in range(500):
+            try:
+                return await self.headerdb.coro_get_block_header_by_hash(parent_hash)
+            except KeyError:
+                self.wait(asyncio.sleep(0.05))
```

Comment:
Why would we get a KeyError and not a HeaderNotFound here? Also, in which situation is the retrying necessary?

Reply:
Hm, for some reason I thought we might receive headers out of order, like:

```
Peer A: request 1-5
Peer B: request 6-10
Peer B: respond header 6-10
```

Anyway, that doesn't appear to be true (at least for now) so I can remove this whole method.